src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,950 @@
+//!   and try to acquire the lock again. No delayed waking up is necessary.
+//!   **Note that though the `LockWaitQueues` is designed to accept it, this
+//!   kind of requests are currently not implemented yet.**
+//!
+//! ## Details about delayed waking up
+//!
+//! The delayed waking-up is needed after waking up a request in legacy mode.
+//! The reasons are:
+//!
+//! * The head of the queue (let's denote its belonging transaction by `T1`) is
+//!   woken up after the current lock being released, then the request will
+//!   return a `WriteConflict` error immediately, and the key is left unlocked.
+//!   It's possible that `T1` won't lock the key again. However, the other
+//!   waiting requests need releasing-lock event to be woken up. In this case,
+//!   we should not let them wait forever until timeout.
+//! * When many transactions are blocked on the same key, and a transaction is
+//!   granted the lock after another releasing the lock, the transaction that's
+//!   blocking other transactions is changed. After cancelling the other
+//!   transactions and let them retry the `AcquirePessimisticLock` request, they
+//!   will be able to re-detect deadlocks with the latest information.
+//!
+//! To achieve this, after delaying for `wake-up-delay-duration` since the
+//! latest waking-up event on the key, a call to
+//! [`LockWaitQueues::delayed_notify_all`] will be made. However, since the
+//! [`LockWaitQueues`] do not have its own thread pool, the user may receive a
+//! future after calling some of the functions, and the user will be responsible
+//! for executing the future in a suitable place.
+
+use std::{
+    collections::BinaryHeap,
+    future::Future,
+    pin::Pin,
+    result::Result,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use dashmap;
+use futures_util::compat::Future01CompatExt;
+use kvproto::kvrpcpb;
+use smallvec::SmallVec;
+use sync_wrapper::SyncWrapper;
+use tikv_util::{time::InstantExt, timer::GLOBAL_TIMER_HANDLE};
+use txn_types::{Key, TimeStamp};
+
+use crate::storage::{
+    errors::SharedError,
+    lock_manager::{lock_wait_context::LockWaitContextSharedState, LockManager, LockWaitToken},
+    mvcc::{Error as MvccError, ErrorInner as MvccErrorInner},
+    txn::Error as TxnError,
+    types::{PessimisticLockParameters, PessimisticLockRes},
+    Error as StorageError,
+};
+
+pub type CallbackWithSharedError<T> = Box<dyn FnOnce(Result<T, SharedError>) + Send + 'static>;
+pub type PessimisticLockKeyCallback = CallbackWithSharedError<PessimisticLockRes>;
+
+/// Represents an `AcquirePessimisticLock` request that's waiting for a lock,
+/// and contains the request's parameters.
+pub struct LockWaitEntry {
+    pub key: Key,
+    pub lock_hash: u64,
+    // TODO: Use term to filter out stale entries in the queue.
+    // pub term: Option<NonZeroU64>,
+    pub parameters: PessimisticLockParameters,
+    pub lock_wait_token: LockWaitToken,
```

Comment:
When will `lock_wait_token` in `LockWaitEntry` be used? I find it also in `LockWaitContextSharedState`, and find `lock_wait_token` here never used...

Reply:
The field in `LockWaitEntry` will be used in the future for updating lock waiting info to the `WaiterManager`, to provide up-to-time diagnostic info and redo deadlock detection when the current-holding-lock is changed.
The field in `LockWaitContextSharedState` will be used in function `finish_request` to remove the waiter from the `WaiterManager`, but commented out since `WaiterManager` is not yet adapted.
The `LockWaitToken` type is used to uniquely identify a lock-waiting request in `WaiterManager`, which is not changed in this PR. The related changes are in #13447 .